### PR TITLE
Bugfix/handle summer time in calendar

### PR DESCRIPTION
**Pull request by adrianmroz**

### CODE REVIEW

**Comment by erykpiast** [comment]
https://www.youtube.com/watch?v=wvUQcnfwUUM

File: src/client/components/date-range-picker/calendar.mocha.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    expect(a).to.be.deep.equal(b);
+  }
+}
+
+const moment = (day: string, timezone: Timezone) => tz(day, timezone.toString());
+const date = (day: string, timezone: Timezone) => moment(day, timezone).toDate();
+
+describe("monthToWeeks", () => {
+
+  describe("March 2010", () => {
```
Comment (erykpiast) [inline src/client/components/date-range-picker/calendar.mocha.ts]: Why is March so special? Describe it.

File: src/client/components/date-range-picker/calendar.mocha.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+});
+
+describe("shiftOneDay", () => {
+  const utc = Timezone.UTC;
+  const warsaw = Timezone.fromJS("Europe/Warsaw");
+
+  it("shift by one day in UTC", () => {
+    const start = moment("2010-01-03", utc);
+    const expected = moment("2010-01-04", utc);
+    expect(shiftOneDay(start).isSame(expected)).to.be.true;
```
Comment (erykpiast) [inline src/client/components/date-range-picker/calendar.mocha.ts]: Did you look for Moment matchers for Chai? `expect(shiftOneDay(start)).to.be.sameDate(expected)` would be so cool.

File: src/client/components/date-range-picker/calendar.mocha.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  });
+
+  it("shift by one day in Europe/Warsaw", () => {
+    const start = moment("2010-01-03", warsaw);
+    const expected = moment("2010-01-04", warsaw);
+    expect(shiftOneDay(start).isSame(expected)).to.be.true;
+  });
+
+  it("shift by one day in Europe/Warsaw across DST forwards", () => {
```
Comment (erykpiast) [inline src/client/components/date-range-picker/calendar.mocha.ts]: wrap those two in `describe("shift by one day in Europe/Warsaw across DST forwards")` and entitle them as `the day of time change` and `the day after time change` or something like this.

File: src/client/components/date-range-picker/calendar.mocha.ts (hunk 1 above)
Comment (adrianmroz) [reply src/client/components/date-range-picker/calendar.mocha.ts]: DST stuff, lol

File: src/common/utils/functional/functional.ts [hunk 4]
```diff
@@ -68,6 +68,16 @@ export function complement<T>(p: Predicate<T>): Predicate<T> {
   return (x: T) => !p(x);
 }
 
+export function range(from: number, to: number): number[] {
+  const result = [];
```
[the remark this answers is not in the document]
Comment (adrianmroz) [reply src/common/utils/functional/functional.ts]: I think it’s too clever. @mkuthan, what your non-js mind thinks?

File: src/client/components/date-range-picker/calendar.mocha.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+  }
+}
+
+const moment = (day: string, timezone: Timezone) => tz(day, timezone.toString());
+const date = (day: string, timezone: Timezone) => moment(day, timezone).toDate();
+
+describe("monthToWeeks", () => {
+
+  describe("March 2010", () => {
+    const march2010Weeks = (tz: Timezone) => [
```
[the remark this answers is not in the document]
Comment (mkuthan) [reply src/client/components/date-range-picker/calendar.mocha.ts]: yep, a lot of duplicated code like  date("2010-03-02", tz), I would make a conversion from "2019-03-02" to date in assertEqualCalendarMatrix method once.

File: src/client/components/date-range-picker/calendar.mocha.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      [date("2010-03-14", utc), date("2010-03-15", utc), date("2010-03-16", utc), date("2010-03-17", utc), date("2010-03-18", utc), date("2010-03-19", utc), date("2010-03-20", utc)],
+      [date("2010-03-21", utc), date("2010-03-22", utc), date("2010-03-23", utc), date("2010-03-24", utc), date("2010-03-25", utc), date("2010-03-26", utc), date("2010-03-27", utc)],
+      [date("2010-03-28", utc), date("2010-03-29", utc), date("2010-03-30", utc), date("2010-03-31", utc), date("2010-04-01", utc), date("2010-04-02", utc), date("2010-04-03", utc)]
+    ]);
+  });
+});
+
+describe("shiftOneDay", () => {
+  const utc = Timezone.UTC;
+  const warsaw = Timezone.fromJS("Europe/Warsaw");
```
Comment (mkuthan) [inline src/client/components/date-range-picker/calendar.mocha.ts]: duplicated in many tests

File: src/client/components/date-range-picker/calendar.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { day, Timezone } from "chronoshift";
+import { Moment } from "moment-timezone/moment-timezone";
+import { range } from "../../../common/utils/functional/functional";
+import { getMoment, Locale } from "../../../common/utils/time/time";
+
+export function calendarDays(startDay: Date, timezone: Timezone, locale: Locale): Date[][] {
```
Comment (mkuthan) [inline src/client/components/date-range-picker/calendar.ts]: What is a difference between calendarDays and monthToWeeks functions? Only padding for first/last weeks? I'm not sure why the method names are so different.

File: src/client/components/date-range-picker/date-range-picker.tsx [hunk 8]
```diff
@@ -154,13 +155,7 @@ export class DateRangePicker extends React.Component<DateRangePickerProps, DateR
...
   renderCalendar(startDate: Date): JSX.Element[] {
     const { timezone } = this.props;
-    const weeks: Date[][] = monthToWeeks(startDate, timezone, getLocale());
-    const firstWeek = weeks[0];
-    const lastWeek = weeks[weeks.length - 1];
-    const countPrepend = 7 - firstWeek.length;
-    const countAppend = 7 - lastWeek.length;
-    weeks[0] = prependDays(timezone, firstWeek, countPrepend);
-    weeks[weeks.length - 1] = appendDays(timezone, lastWeek, countAppend);
+    const weeks: Date[][] = calendarDays(startDate, timezone, getLocale());
```
Comment (mkuthan) [inline src/client/components/date-range-picker/date-range-picker.tsx]: I really like this change, logic not in the react component but in separate functions.

File: src/common/utils/functional/functional.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkuthan) [reply src/common/utils/functional/functional.ts]: https://www.scala-lang.org/api/2.12.2/scala/collection/immutable/Range.html :P

[the remark this answers is not in the document]
Comment (adrianmroz) [reply src/common/utils/functional/functional.ts]: `sealed class Range extends AbstractSeq[Int] with IndexedSeq[Int] with CustomParallelizable[Int, ParRange] with Serializable` cool stuff

File: src/client/components/date-range-picker/calendar.mocha.ts (hunk 6 above)
Comment (adrianmroz) [reply src/client/components/date-range-picker/calendar.mocha.ts]: promoted to global constants

File: src/client/components/date-range-picker/calendar.ts (hunk 7 above)
Comment (adrianmroz) [reply src/client/components/date-range-picker/calendar.ts]: Yes. monthToWeeks produces array of dates from month grouped by weeks. calendarDays additionally pads with dates from previous/next months. Any naming ideas? `paddedWeeks`? :o

Comment (mkuthan) [reply src/client/components/date-range-picker/calendar.ts]: monthToPaddedWeeks to be consistent, or single function with parameter padded=true|false

Comment (adrianmroz) [reply src/client/components/date-range-picker/calendar.ts]: Not sure if it’s clear, but only calendarDay is public API. monthToWeeks is internal function. First you need to get month to weeks, then you pad around do fill space on calendar sheet and return that.

Comment (mkuthan) [reply src/client/components/date-range-picker/calendar.ts]: ok  if public api is consistent

**Pull request merged**